sstables/compaction_manager.cc

```diff
@@ -596,39 +596,60 @@ future<> compaction_manager::rewrite_sstables(column_family* cf, sstables::compa
     task->compacting_cf = cf;
     task->type = options.type();
     _tasks.push_back(task);
-    _stats.pending_tasks++;
 
-    task->compaction_done = repeat([this, task, options, get_func = std::move(get_func)] () mutable {
+    struct work {
+        size_t total_pending;
+        std::vector<sstables::shared_sstable> sstables;
+        work(std::vector<sstables::shared_sstable> sstables)
+            : total_pending(sstables.size())
+            , sstables(std::move(sstables))
+        {}
+    };
+
+    auto wrk = std::make_unique<work>(get_func(*cf));
+    auto w = wrk.get();
+    _stats.pending_tasks += w->total_pending;
+
+    task->compaction_done = do_for_each(w->sstables.begin(), w->sstables.end(),
+            [this, task, options, w] (sstables::shared_sstable sst) mutable {
+
         // FIXME: lock cf here
         if (!can_proceed(task)) {
-            _stats.pending_tasks--;
-            return make_ready_future<stop_iteration>(stop_iteration::yes);
+            return make_ready_future<>();
         }
+
+        return repeat([this, task, options, sst, w] () mutable {
         column_family& cf = *task->compacting_cf;
-        sstables::compaction_descriptor descriptor = sstables::compaction_descriptor(get_func(cf));
-        descriptor.options = options;
+        auto sstable_level = sst->get_sstable_level();
+        auto run_identifier = sst->run_identifier();
+        auto descriptor = sstables::compaction_descriptor({ sst }, sstable_level,
+                    sstables::compaction_descriptor::default_max_sstable_bytes, run_identifier, options);
+
         auto compacting = make_lw_shared<compacting_sstable_registration>(this, descriptor.sstables);
         // Releases reference to cleaned sstable such that respective used disk space can be freed.
         descriptor.release_exhausted = [compacting] (const std::vector<sstables::shared_sstable>& exhausted_sstables) {
             compacting->release_compacting(exhausted_sstables);
         };
 
+        w->total_pending--;
         _stats.pending_tasks--;
         _stats.active_tasks++;
         task->compaction_running = true;
+
         compaction_backlog_tracker user_initiated(std::make_unique<user_initiated_backlog_tracker>(_compaction_controller.backlog_of_shares(200), _available_memory));
         return do_with(std::move(user_initiated), [this, &cf, descriptor = std::move(descriptor)] (compaction_backlog_tracker& bt) mutable {
             return with_scheduling_group(_scheduling_group, [this, &cf, descriptor = std::move(descriptor)] () mutable {
-                return cf.rewrite_sstables(std::move(descriptor));
+                return cf.run_compaction(std::move(descriptor));
```

Comment:
It's confusing that `table` has both `run_compaction()` and `compact_sstables()`, the former calling the latter. Not relevant to this patch.

Reply:
we'll get there.